quodlibet/quodlibet/ext/events/synchronize_to_device.py

```diff
@@ -0,0 +1,144 @@
+import os
+import string
+from threading import Thread
+
+from gi.repository import Gtk, GLib
+from pydub import AudioSegment
+from quodlibet import _, app
+from quodlibet import config
+from quodlibet import get_user_dir
+from quodlibet import qltk
+from quodlibet.compat import itervalues
+from quodlibet.plugins import PluginConfigMixin
+from quodlibet.plugins.events import EventPlugin
+from quodlibet.qltk.ccb import ConfigCheckButton
+from quodlibet.query import Query
+
+
+class SyncToDevice(EventPlugin, PluginConfigMixin):
+    PLUGIN_ID = "synchronize_to_device"
+    PLUGIN_NAME = _("Synchronize to device")
+    PLUGIN_DESC = _(
+        "Converts all songs from the selected saved searches to 192 kbs and synchronizes them with the specified" 
+        "folder. All songs in that folder, which are not in the saved searches, will be deleted."
+        "Needs the python package 'pydub'.")
+    c_path = __name__ + '_path'
+
+    def PluginPreferences(self, parent):
+        vbox = Gtk.VBox(spacing=6)
+
+        queries = {}
+        query_path = os.path.join(get_user_dir(), 'lists', 'queries.saved')
+        try:
+            with open(query_path, 'rU') as query_file:
+                log = Gtk.TextView()
+                log.set_left_margin(5)
+                log.set_right_margin(5)
+                log.editable = False
+                buffer = log.get_buffer()
+                scroll = Gtk.ScrolledWindow()
+                scroll.set_size_request(-1, 100)
+                scroll.add(log)
+
+                def append(text):
+                    GLib.idle_add(lambda: buffer.insert(buffer.get_end_iter(), text + "\n"))
+                    GLib.idle_add(lambda: log.scroll_to_mark(buffer.get_insert(), 0.0, True, 0.5, 0.5))
+
+                for query_string in query_file:
+                    name = next(query_file).strip()
+                    queries[name] = Query(query_string.strip())
+
+                for query_name, query in queries.items():
+                    check_button = ConfigCheckButton(query_name, "plugins", self._config_key(query_name))
+                    check_button.set_active(self.config_get_bool(query_name))
+                    vbox.pack_start(check_button, False, True, 0)
+
+                def synchronize():
```

Comment:
All the quodlibet API isn't thread safe (except if explicitly stated), so threads wont work here.

Reply:
Thanks for the excellent code review!
Without using threads the UI will become unresponsive until the synchronisation has finished.
What could happen due to the not-thread-safe API? I think the worst scenario are wrong songs on the device.
The user is aware of the ongoing sync and won't edit the library.
So I propose to keep the threads.

Reply:
So what shall we do here?